src/replication.c

```diff
@@ -1112,15 +1113,65 @@ void restartAOFAfterSYNC() {
     }
 }
 
+long long getMonotonicTimeMs() {
+    struct timespec res;
+    clock_gettime(CLOCK_MONOTONIC, &res);
+    return (res.tv_sec * 1000000000LL + res.tv_nsec) / 1000000LL;
+}
+
+#define SLAVE_SYNC_BUDGET_UPDATE_INTERVAL_MS 10
+#define SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS 50
+void updateSlaveSyncBudget(ssize_t nread) {
+    if (server.slave_sync_speed_bps <= 0) return;
+    server.slave_sync_budget -= nread * 8000LL;
+
+    long long now_ms = getMonotonicTimeMs();
+
+    long long delta_time_ms = now_ms - server.slave_sync_budget_time_ms;
+    if (delta_time_ms > SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS) {
+        delta_time_ms = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS;
+    }
+    long long max_budget = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS * server.slave_sync_speed_bps;
```

Comment:
Making a note in case anyone else has the same question - `long long` is guaranteed to be at least 64 bits, so this should be large enough - e.g. transmitting a terabyte of data in 10ms (extremely unrealistic, internet is closer to 10 gigabytes/second) is `1<<43` bits, or 1<<53 millibits (1000 =~ 1<<10)